chctools/btor2.py

```diff
@@ -870,26 +874,95 @@ def inv(self) -> z3.FuncDeclRef:
-                "(exit)\n"]
+    def bad(self) -> z3.ExprRef:
+        self.reduce_bads()
+        return self.bads[0]
+
+    def reduce_bads(self) -> None:
+        if len(self.bads) > 1:
+            self.bads = [z3.Or(self.bads)]
+
+    def reduce_constraints(self, method: int = 0) -> None:
+        if self.constraints:
+            # (init(v), tr(v, in, v'), bad(v)) with constraint c(v) is equiv to
+            # init(v) and c(v), tr(v, in, v') and c(v), bad(v) and c(v)
+            # (init(v), tr(v, in, v'), bad(v)) with constraint c(v, in) is equiv to
+            # init(v), tr(v, in, v') and c(v, in), bad(v)
+            if method == 0:
+                self.init = z3.And(self.init, *self.constraints)
+                self.tr = z3.And(self.tr, *self.constraints)
+                self.bads = [z3.And(self.bad(), *self.constraints)]
+            elif method == 1:
+                self.tr = z3.And(self.tr, *self.constraints)
+            else:
+                raise NotImplementedError
+            self.constraints = []
+
+
+def generate_vc(ts: Ts) -> Tuple[z3.ExprRef, z3.ExprRef, z3.ExprRef, z3.FuncDeclRef]:
+    if len(ts.bads) > 1 or ts.constraints:
+        raise NotImplementedError
+    inputs_and_vars: List[z3.ExprRef] = ts.vars_and_inputs()
+    inv: z3.FuncDeclRef = ts.inv()
+    inv_pre: z3.ExprRef = inv(*ts.pre_vars)
+    inv_post: z3.ExprRef = inv(*ts.post_vars)
+    return (z3.ForAll(inputs_and_vars, z3.Implies(ts.init, inv_pre)),
+            z3.ForAll(inputs_and_vars, z3.Implies(z3.And(inv_pre, ts.tr), inv_post)),
+            z3.ForAll(inputs_and_vars, z3.Implies(z3.And(inv_pre, ts.bad()), False)),
+            inv)
+
+
+def generate_chc_str(init: z3.ExprRef, tr: z3.ExprRef, bad: z3.ExprRef,
+                     inv: z3.FuncDeclRef) -> List[str]:
+    return ["(set-logic HORN)\n",
+            "(set-option :fp.engine spacer)\n",
+            "{:s}\n".format(inv.sexpr()),
+            "(assert {:s})\n".format(init.sexpr()),
+            "(assert {:s})\n".format(tr.sexpr()),
+            "(assert {:s})\n".format(bad.sexpr()),
+            "(check-sat)\n",
+            "(get-model)\n",
+            "(exit)\n"]
+
+
+def generate_horn_clause_db(init: z3.ExprRef, tr: z3.ExprRef, bad: z3.ExprRef) -> HornClauseDb:
+    db: HornClauseDb = HornClauseDb()
+    db.add_rule(HornRule(init))
+    db.add_rule(HornRule(tr))
+    db.add_rule(HornRule(bad))
+    db.seal()
+    return db
+
+
+def bmc(init: z3.ExprRef, tr: z3.ExprRef, bad: z3.ExprRef, n: int, pre_vars: List[z3.ExprRef],
+        post_vars: List[z3.ExprRef], inputs: List[z3.ExprRef]):
+    z3_exprs: List[z3.ExprRef] = [init, tr]
+
+    vars1: List[z3.ExprRef] = pre_vars.copy()
+    vars2: List[z3.ExprRef] = post_vars.copy()
+    ips: List[z3.ExprRef] = inputs.copy()
+
+    for i in range(n - 1):
+        for j in range(len(vars1)):
+            var: z3.ExprRef = z3.FreshConst(vars2[j].sort(), vars2[j].sexpr().split('!')[0])
+            tr = z3.substitute(tr, (vars2[j], var))
```

Comment:
dont call substitute in a loop. instead create a list of substitutions and call it ones

Reply:
I just realized that substitutes can take an arbitrary number of pairs. It is fixed in the new PR.